test/mechanics/integrands/HistoryData.cpp

```diff
@@ -0,0 +1,397 @@
+#include "mechanics/interpolation/InterpolationTrussLinear.h"
+#include "mechanics/mesh/MeshFem.h"
+#include "mechanics/mesh/MeshFemDofConvert.h"
+#include "mechanics/nodes/NodeSimple.h"
+
+#include <cassert>
+#include <functional>
+#include <map>
+#include <vector>
+
+using namespace Eigen;
+using namespace NuTo;
+using namespace NuTo::Groups;
+
+
+constexpr double SpecimenLength = 10.;
+constexpr unsigned int numElements = 4;
+
+//! @brief Standard history data management object.
+//! @tparam T: Data type of the history data
+template <typename T>
+class HistoryData
+{
+    std::map<int, std::vector<T>> mHisData;
+
+public:
+    //! @brief Returns the history data for a specific integration point
+    //! @param cellNum: Number of the cell containing the integration point
+    //! @param ipNum: Number of the integration point
+    //! @return History data of the integration point
+    const T& GetIPHistoryData(int cellNum, int ipNum) const
+    {
+        auto ipVecIt = mHisData.find(cellNum);
+        if (ipVecIt == mHisData.end())
+            throw Exception(__PRETTY_FUNCTION__, "Cell data not found! - Did you initialize the history data?");
+        return (*ipVecIt).second[ipNum];
+    }
+
+    //! @brief Returns the history data for a specific integration point
+    //! @param cellNum: Number of the cell containing the integration point
+    //! @param ipNum: Number of the integration point
+    //! @return History data of the integration point
+    T& GetIPHistoryData(int cellNum, int ipNum)
+    {
+        auto ipVecIt = mHisData.find(cellNum);
+        if (ipVecIt == mHisData.end())
+            throw Exception(__PRETTY_FUNCTION__, "Cell data not found! - Did you initialize the history data?");
+        return (*ipVecIt).second[ipNum];
+    }
+
+    //! @brief Initializes the history data
+    //! @param cells: Cellgroup that is related to the integrand
+    //! @param dof: Dof type
+    void InitializeHistoryData(const Groups::Group<CellInterface>& cells, DofType dof)
+    {
+        for (CellInterface& cell : cells)
+        {
+            cell.Apply([this](const NuTo::CellData& cellData, const NuTo::CellIpData& cellIpData) {
+                if (mHisData.find(cellData.GetCellId()) == mHisData.end())
+                    mHisData.emplace(cellData.GetCellId(), std::vector<T>(cellData.GetNumIntegrationPoints()));
+            });
+        }
+    }
+};
+
+
+// %%% Custom law with history data %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+//! @brief Structure that holds the history data of the creep law
+struct CreepHistoryData
+{
+    EngineeringStrainPDE<1> prevStrain;
+    EngineeringStressPDE<1> prevStress;
+    VectorXd gamma = Eigen::VectorXd::Zero(2);
+
+    CreepHistoryData()
+    {
+        prevStrain[0] = 0.;
+        prevStress[0] = 0.;
+    }
+};
+
+//! @brief Creep law using the exponential algorithm
+class CreepLaw : public Laws::MechanicsInterface<1>, public HistoryData<CreepHistoryData>
```

Comment:
I would prefer composition over inheritance. But this is no `Core` code...

Reply:
Member is also okay, but than you have to to write a pass through for the Initialize function or a getter to the history data. I think inheritance is the easier way. YMMV